api/client/src/main/java/org/projectnessie/client/auth/oauth/OAuth2Client.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.client.auth.oauth;
+
+import static org.projectnessie.client.auth.oauth.OAuth2ClientParams.MIN_REFRESH_DELAY;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.URI;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import org.projectnessie.client.http.HttpClient;
+import org.projectnessie.client.http.HttpClientException;
+import org.projectnessie.client.http.HttpRequest;
+import org.projectnessie.client.http.HttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A simple OAuth2 client that supports the "client_credentials" grant type for fetching new access
+ * tokens, using pre-defined client credentials.
+ *
+ * <p>This client also supports refreshing access tokens using both the refresh token method defined
+ * in RFC 6749 and the token exchange method defined in RFC 8693.
+ *
+ * <p>If you don't need to refresh access tokens, you can use the {@link
+ * org.projectnessie.client.auth.BearerAuthenticationProvider BearerAuthenticationProvider} instead
+ * and provide the token to use directly through configuration.
+ */
+public class OAuth2Client implements OAuth2Authenticator, Closeable {
```

Comment:
I suspect we should let `NessieAuthenticator` extend `AutoCloseable` w/ a default noop and close associated authenticators when the client's closed

Reply:
That was my first intention, but then I backed off because the changes are a bit invasive: you need to "bubble up" the call to `close()` up to all `NessieApi` impls (they are already `AutoCloseable`, but nothing was enforcing that). But if you are OK with that, let's do it.

Reply:
Yea, `NessieApi` instances should always be properly closed from call sites. I'm (mostly) sure, that we always do that.

Reply:
(Except that Iceberg `Catalog`s do not implement `AutoCloseable`, but some implementations do *sigh*)